Objects/listobject.c

```diff
@@ -590,37 +592,60 @@ list_ass_slice(PyListObject *a, Py_ssize_t ilow, Py_ssize_t ihigh, PyObject *v)
     Py_ssize_t k;
     size_t s;
     int result = -1;            /* guilty until proved innocent */
+
+    if (ilow < 0)
+        ilow = 0;
+    else if (ilow >= Py_SIZE(a))
+        ilow = Py_SIZE(a);
+
+    if (ihigh < ilow)
+        ihigh = ilow;
+    else if (ihigh > Py_SIZE(a))
+        ihigh = Py_SIZE(a);
+
+    norig = ihigh - ilow;
+    assert(norig >= 0);
 #define b ((PyListObject *)v)
     if (v == NULL)
         n = 0;
     else {
         if (a == b) {
-            /* Special case "a[i:j] = a" -- copy b first */
-            v = list_slice(b, 0, Py_SIZE(b));
-            if (v == NULL)
-                return result;
-            result = list_ass_slice(a, ilow, ihigh, v);
-            Py_DECREF(v);
-            return result;
+            /* Special case "a[i:j] = a" -- no need for DECREFs */
+            if (norig == Py_SIZE(a)) {
+                return 0;
+            }
+            Py_ssize_t orig_size = Py_SIZE(a);
+            if (list_resize(a, orig_size + orig_size - norig) != 0) {
+                return -1;
+            }
+            for (k = 0; k < orig_size - ihigh; k++) {
+                PyObject *obj = a->ob_item[ihigh + k];
+                Py_INCREF(obj);
+                a->ob_item[orig_size + ilow + k] = obj;
+            }
+            memmove(a->ob_item + ilow + ilow, a->ob_item + ilow, (orig_size - ilow) * sizeof(PyObject*));
+            for (k = 0; k < ilow; k++) {
+                PyObject *obj = a->ob_item[k];
+                Py_INCREF(obj);
+                a->ob_item[k + ilow] = obj;
```

Comment:
Nitpick: I prefer `a->ob_item[ilow + k] = obj;` as I always tend to read this as `point_of_origin + moving_index` . Also is consistent with the rest of the movements.